File: tools/patcher.py
from __future__ import annotations

import subprocess
from functools import cache
from pathlib import Path

import lief

from .compiler import extract_cave_asm

NOP = b"\x1f\x20\x03\xd5"
DST = b"\xbe\xba\xfe\xca"
NEXT = b"\xfe\xca\xad\xde"
CONV = b"\xfe\xca\xef\xbe"


def encode_b(src_va: int, dst_va: int) -> int:
    delta = dst_va - src_va
    if delta % 4:
        raise ValueError("branch target must be 4-byte aligned")
    imm = delta >> 2
    if not -(1 << 25) <= imm < (1 << 25):
        raise ValueError("branch target out of range")
    return 0x14000000 | (imm & 0x03FFFFFF)


def encode_bl(src_va: int, dst_va: int) -> int:
    delta = dst_va - src_va
    if delta % 4:
        raise ValueError("branch target must be 4-byte aligned")
    imm = delta >> 2
    if not -(1 << 25) <= imm < (1 << 25):
        raise ValueError("branch target out of range")
    return 0x94000000 | (imm & 0x03FFFFFF)


def _branch(src_va: int, dst_va: int, scratch: int = 16) -> bytes:
    try:
        return encode_b(src_va, dst_va).to_bytes(4, "little")
    except ValueError:
        src_page = src_va & ~0xFFF
        dst_page = dst_va & ~0xFFF
        imm21 = ((dst_page - src_page) >> 12) & 0x1FFFFF
        adrp = 0x90000000 | scratch | ((imm21 & 3) << 29) | (((imm21 >> 2) & 0x7FFFF) << 5)
        add = 0x91000000 | scratch | (scratch << 5) | ((dst_va & 0xFFF) << 10)
        br = 0xD61F0000 | (scratch << 5)
        return adrp.to_bytes(4, "little") + add.to_bytes(4, "little") + br.to_bytes(4, "little")


def _target(insn: int, va: int) -> int | None:
    if (insn & 0xFC000000) in (0x14000000, 0x94000000):
        imm = insn & 0x3FFFFFF
        if imm & 0x2000000:
            imm -= 0x4000000
        return va + imm * 4
    if (insn & 0xFF000010) == 0x54000000 or (insn & 0x7E000000) == 0x34000000:
        imm = (insn >> 5) & 0x7FFFF
        if imm & 0x40000:
            imm -= 0x80000
        return va + imm * 4
    if (insn & 0x7E000000) == 0x36000000:
        imm = (insn >> 19) & 0x3FFF
        if imm & 0x2000:
            imm -= 0x4000
        return va + imm * 4
    return None


def _patched(original: bytes, hook_va: int, cave_va: int) -> bytes:
    out = bytearray()
    cur = cave_va
    for i in range(0, len(original), 4):
        raw = original[i:i + 4]
        insn = int.from_bytes(raw, "little")
        va = hook_va + i
        if (insn & 0xFC000000) == 0x14000000:
            b = _branch(cur, _target(insn, va))
            out += b
            cur += len(b)
        elif (insn & 0xFC000000) == 0x94000000:
            out += encode_bl(cur, _target(insn, va)).to_bytes(4, "little")
            cur += 4
        else:
            out += raw
            cur += 4
    return bytes(out)


def _patched_size(original: bytes) -> int:
    return len(_patched(original, 0, 0))


def _reg(name: str) -> tuple[int, bool]:
    name = name.lower().strip()
    return int(name[1:]), name.startswith("x")


def _mov(rd: int, rm: int, is64: bool) -> bytes:
    insn = (0xAA0003E0 if is64 else 0x2A0003E0) | (rm << 16) | rd
    return insn.to_bytes(4, "little")


def _wrapper(regs: list[str], wrapper_va: int, plugin_va: int) -> bytes:
    out = bytearray()
    for i, r in enumerate(regs[:8]):
        n, is64 = _reg(r)
        out += _mov(i, n, is64)
    out += encode_b(wrapper_va + len(out), plugin_va).to_bytes(4, "little")
    return bytes(out)


@cache
def _frame() -> tuple[bytes, bytes, bytes]:
    return extract_cave_asm()


def build_hook_cave(cave_va: int, hook_va: int, hook_size: int, original: bytes, plugin_blobs: list, target_binary: Path | None = None, detour: bool = False, branch_host: bool = False, nop_addrs: list[int] | None = None) -> bytes:
    if not original or len(original) % 4 or hook_size % 4:
        raise ValueError("invalid hook window")
    from .compiler import PluginBlob
    save, restore, ret = _frame()
    n = len(plugin_blobs)
    target_dst = None
    if branch_host:
        detour = True
        for i in range(0, len(original), 4):
            target_dst = _target(int.from_bytes(original[i:i + 4], "little"), hook_va + i)
            if target_dst is not None:
                break
        if target_dst is None:
            raise ValueError("branch hook needs a branch instruction")
    patched = original if detour else _patched(original, hook_va, cave_va + 4 + n * 4 + 4)
    control_size = 4 + n * 4 + 4 + (4 if detour else len(patched) + 4)
    offsets, chunks = [], []
    cur = control_size
    for blob in plugin_blobs:
        cur = (cur + 3) & ~3
        regs = blob.register_args if isinstance(blob, PluginBlob) else None
        entry = blob.entry_offset if isinstance(blob, PluginBlob) else 0
        if regs:
            woff = cur
            cur += 4 * len(regs) + 4
            cur = (cur + 3) & ~3
            poff = cur
            offsets.append(woff)
            built = blob.build(cave_va + poff, cave_va + poff + ((len(blob.text) + 15) & ~15), target_binary)
            chunks.append(_wrapper(regs, cave_va + woff, cave_va + poff + entry))
            chunks.append(built)
            cur = poff + len(built)
        else:
            offsets.append(cur + entry)
            if isinstance(blob, PluginBlob):
                built = blob.build(cave_va + cur, cave_va + cur + ((len(blob.text) + 15) & ~15), target_binary)
            else:
                built = blob
            chunks.append(built)
            cur += len(built)
    out = bytearray(save)
    for i, off in enumerate(offsets):
        out += encode_bl(cave_va + 4 + i * 4, cave_va + off).to_bytes(4, "little")
    out += restore
    out += ret if detour else patched + _branch(cave_va + 4 + n * 4 + 4 + len(patched), hook_va + hook_size)
    for chunk in chunks:
        out += chunk
        out += b"\x00" * ((-len(out)) % 4)
    if target_dst is not None:
        for marker, target in ((DST, target_dst), (NEXT, hook_va + 4), (CONV, hook_va + hook_size)):
            idx = 0
            while True:
                idx = out.find(marker, idx)
                if idx < 0:
                    break
                out[idx:idx + 4] = encode_b(cave_va + idx, target).to_bytes(4, "little")
                idx += 4
    return bytes(out)


def _codesign(path: Path) -> None:
    subprocess.run(["codesign", "--force", "--sign", "-", str(path)], capture_output=True)


def patch_hook_window(binary_path: Path, output_path: Path, src_va: int, size: int, dst_va: int) -> None:
    binary = lief.parse(str(binary_path))
    if binary is None:
        raise RuntimeError(f"failed to parse {binary_path}")
    off = binary.virtual_address_to_offset(src_va)
    if not isinstance(off, int) or off < 0:
        raise ValueError(f"cannot map VA 0x{src_va:x}")
    hook = _branch(src_va, dst_va)
    if len(hook) > size:
        raise ValueError(f"hook window 0x{size:x} too small")
    data = bytearray(Path(output_path).read_bytes())
    data[off:off + len(hook)] = hook
    for pos in range(len(hook), size, 4):
        data[off + pos:off + pos + 4] = NOP
    Path(output_path).write_bytes(data)


def patch_bytes_va(binary_path: Path, output_path: Path, va: int, payload: bytes) -> None:
    binary = lief.parse(str(binary_path))
    if binary is None:
        raise RuntimeError(f"failed to parse {binary_path}")
    off = binary.virtual_address_to_offset(va)
    if not isinstance(off, int) or off < 0:
        raise ValueError(f"cannot map VA 0x{va:x}")
    data = bytearray(Path(output_path).read_bytes())
    if off + len(payload) > len(data):
        raise ValueError(f"patch out of range: 0x{va:x}")
    data[off:off + len(payload)] = payload
    Path(output_path).write_bytes(data)


def patch_hook_macho(binary_path: Path, output_path: Path, hook_file_off: int, hook_size: int, original: bytes, plugin_blobs: list, seg_name: str, detour: bool = False, branch_host: bool = False, nop_addrs: list[int] | None = None) -> tuple[int, int]:
    from .segment import SegmentPlan, add_segment, remap_macho_offset_va, seg_va, segment_file_offset, write_at_offset
    binary = lief.parse(str(binary_path))
    if binary is None or not isinstance(binary, lief.MachO.Binary):
        raise RuntimeError(f"failed to parse {binary_path}")
    control = 4 + len(plugin_blobs) * 4 + 4 + (4 if detour else _patched_size(original) + 4)
    size = control
    for b in plugin_blobs:
        size += b.total_bytes if hasattr(b, "total_bytes") else len(b)
        if getattr(b, "register_args", None):
            size += 4 * len(b.register_args) + 4
    size = max(size, 4)
    add_segment(output_path, SegmentPlan(seg_name, size, b"\x00" * size), output_path)
    patched_binary = lief.parse(str(output_path))
    if patched_binary is None or not isinstance(patched_binary, lief.MachO.Binary):
        raise RuntimeError(f"failed to parse {output_path}")
    hook_va = remap_macho_offset_va(binary, patched_binary, hook_file_off)
    cave_va = seg_va(patched_binary, seg_name, size)
    cave_off = segment_file_offset(patched_binary, seg_name)
    blob = build_hook_cave(cave_va, hook_va, hook_size, original, plugin_blobs, output_path, detour, branch_host, nop_addrs)
    write_at_offset(output_path, cave_off, blob, max(size, 4))
    patch_hook_window(output_path, output_path, hook_va, hook_size, cave_va)
    _codesign(output_path)
    return hook_va, cave_va

File: src/patcher.cpp
#include "patcher.h"
#include "aarch64/decoder.h"
#include "aarch64/encoder.h"
#include "aarch64/relocator.h"
#include "compiler.h"
#include "segment.h"
#include <cstring>
#include <cstdio>
#include <cstdlib>
#include <algorithm>

static const uint8_t NOP[4] = {0x1f, 0x20, 0x03, 0xd5};
static const uint8_t DST[4] = {0xbe, 0xba, 0xfe, 0xca};
static const uint8_t NEXT[4] = {0xfe, 0xca, 0xad, 0xde};
static const uint8_t CONV[4] = {0xfe, 0xca, 0xef, 0xbe};

uint32_t encode_b(uint64_t src_va, uint64_t dst_va) {
    return armcave::aarch64::encode_b(src_va, dst_va);
}

uint32_t encode_bl(uint64_t src_va, uint64_t dst_va) {
    return armcave::aarch64::encode_bl(src_va, dst_va);
}

uint64_t target_insn(uint32_t insn, uint64_t va) {
    auto decoded = armcave::aarch64::decode(insn, va);
    return decoded.has_target && armcave::aarch64::is_branch(decoded.kind)
        ? decoded.target : 0;
}

static std::vector<uint8_t> patched(const std::vector<uint8_t> &original,
                                     uint64_t hook_va, uint64_t cave_va) {
    return armcave::aarch64::relocate_block(original, hook_va, cave_va).bytes;
}

static int patched_size(const std::vector<uint8_t> &original) {
    return (int)armcave::aarch64::max_relocated_size(original.size());
}

static int reg_num(const std::string &name) {
    return atoi(name.c_str() + 1);
}

static bool reg_is64(const std::string &name) {
    return name[0] == 'x' || name[0] == 'X';
}

static std::vector<uint8_t> mov_insn(int rd, int rm, bool is64) {
    uint32_t insn = (is64 ? 0xAA0003E0 : 0x2A0003E0) | (rm << 16) | rd;
    std::vector<uint8_t> out(4);
    memcpy(out.data(), &insn, 4);
    return out;
}

static std::vector<uint8_t> make_wrapper(const std::vector<std::string> &regs,
                                          uint64_t wrapper_va, uint64_t plugin_va) {
    std::vector<uint8_t> out;
    for (size_t i = 0; i < regs.size() && i < 8; i++) {
        auto m = mov_insn((int)i, reg_num(regs[i]), reg_is64(regs[i]));
        out.insert(out.end(), m.begin(), m.end());
    }
    auto branch = armcave::aarch64::make_branch_sequence(
        wrapper_va + out.size(), plugin_va, false);
    out.insert(out.end(), branch.begin(), branch.end());
    return out;
}

static void append_word(std::vector<uint8_t> &out, uint32_t instruction) {
    out.push_back((uint8_t)(instruction & 0xffU));
    out.push_back((uint8_t)((instruction >> 8) & 0xffU));
    out.push_back((uint8_t)((instruction >> 16) & 0xffU));
    out.push_back((uint8_t)((instruction >> 24) & 0xffU));
}

static std::vector<uint8_t> make_cpp_func_wrapper(
    const std::vector<std::string> &regs,
    uint64_t wrapper_va, uint64_t plugin_va) {
    (void)regs;
    std::vector<uint8_t> out;

    // The call site prepares the first C++ argument in w0. Preserve the
    // caller's LR, then expose the C++ return value in x1.
    append_word(out, 0xa9bf7bfdU); // stp x29, x30, [sp, #-16]!
    auto branch = armcave::aarch64::make_branch_sequence(
        wrapper_va + out.size(), plugin_va, true);
    out.insert(out.end(), branch.begin(), branch.end());
    append_word(out, 0xaa0003e1U); // mov x1, x0
    append_word(out, 0xa8c17bfdU); // ldp x29, x30, [sp], #16
    append_word(out, 0xd65f03c0U); // ret
    return out;
}

static struct CaveFrame {
    std::vector<uint8_t> save;
    std::vector<uint8_t> restore;
    std::vector<uint8_t> ret;
    bool valid = false;
} g_cave_frame;

static void ensure_cave_frame() {
    if (g_cave_frame.valid) return;
    g_cave_frame.save = extract_cave_asm_save();
    g_cave_frame.restore = extract_cave_asm_restore();
    g_cave_frame.ret = extract_cave_asm_ret();
    g_cave_frame.valid = true;
}

int hook_dispatch_size(int handler_count, int original_size, bool override_original,
                       bool strip_pac) {
    ensure_cave_frame();
    int save_size = (int)g_cave_frame.save.size() - (strip_pac ? 0 : 4);
    return save_size + handler_count * (int)armcave::aarch64::kMaxBranchSequenceBytes +
           (int)g_cave_frame.restore.size() +
           (override_original ? (int)g_cave_frame.ret.size() :
            (int)armcave::aarch64::max_relocated_size((size_t)original_size) +
            (int)armcave::aarch64::kMaxBranchSequenceBytes);
}

std::vector<uint8_t> build_hook_dispatch(
    uint64_t cave_va,
    uint64_t hook_va,
    int hook_size,
    const std::vector<uint8_t> &original,
    const std::vector<uint64_t> &handler_vas,
    bool override_original,
    bool strip_pac) {

    if (original.empty() || original.size() % 4 || hook_size % 4)
        throw std::runtime_error("invalid hook window");
    ensure_cave_frame();

    size_t save_offset = strip_pac ? 0 : 4;
    if (g_cave_frame.save.size() < save_offset)
        throw std::runtime_error("invalid cave save sequence");
    uint64_t calls_va = cave_va + g_cave_frame.save.size() - save_offset;
    std::vector<std::vector<uint8_t>> calls;
    uint64_t call_cursor = calls_va;
    for (auto handler_va : handler_vas)
    {
        auto call = armcave::aarch64::make_branch_sequence(
            call_cursor, handler_va, true);
        call_cursor += call.size();
        calls.push_back(std::move(call));
    }
    uint64_t resume_va = call_cursor + g_cave_frame.restore.size();
    std::vector<uint8_t> relocated;
    if (!override_original)
        relocated = patched(original, hook_va, resume_va);

    std::vector<uint8_t> out;
    out.insert(out.end(), g_cave_frame.save.begin() + save_offset,
               g_cave_frame.save.end());
    for (auto &call : calls)
        out.insert(out.end(), call.begin(), call.end());
    out.insert(out.end(), g_cave_frame.restore.begin(), g_cave_frame.restore.end());
    if (override_original) {
        out.insert(out.end(), g_cave_frame.ret.begin(), g_cave_frame.ret.end());
    } else {
        out.insert(out.end(), relocated.begin(), relocated.end());
        auto branch = armcave::aarch64::make_branch_sequence(
            cave_va + out.size(), hook_va + hook_size, false);
        out.insert(out.end(), branch.begin(), branch.end());
    }
    return out;
}

int plugin_wrapper_size(const std::vector<std::string> &registers) {
    return registers.empty() ? 0 : (int)std::min<size_t>(registers.size(), 8) * 4 + 4;
}

int plugin_wrapper_max_size(const std::vector<std::string> &registers) {
    return registers.empty() ? 0 : (int)std::min<size_t>(registers.size(), 8) * 4 +
                                      (int)armcave::aarch64::kMaxBranchSequenceBytes;
}

int cpp_func_wrapper_size(const std::vector<std::string> &registers) {
    return 4 + (int)std::min<size_t>(registers.size(), 8) * 4 + 4 + 12;
}

int cpp_func_wrapper_max_size(const std::vector<std::string> &registers) {
    return 4 + (int)std::min<size_t>(registers.size(), 8) * 4 +
           (int)armcave::aarch64::kMaxBranchSequenceBytes + 12;
}

std::vector<uint8_t> build_plugin_wrapper(
    const std::vector<std::string> &registers,
    uint64_t wrapper_va,
    uint64_t plugin_va,
    bool cpp_func) {
    return cpp_func ? make_cpp_func_wrapper(registers, wrapper_va, plugin_va)
                    : make_wrapper(registers, wrapper_va, plugin_va);
}

int hook_window_size(uint64_t src_va, uint64_t dst_va) {
    return (int)armcave::aarch64::branch_sequence_size(src_va, dst_va, false);
}

std::vector<uint8_t> build_hook_cave(
    uint64_t cave_va, uint64_t hook_va, int hook_size,
    const std::vector<uint8_t> &original,
    const std::vector<PluginBlob *> &plugin_blobs,
    const std::filesystem::path *target_binary,
    bool override_original, bool branch_host,
    const std::vector<uint64_t> *nop_addrs) {
    (void)nop_addrs;
    if (original.empty() || original.size() % 4 || hook_size % 4)
        throw std::runtime_error("invalid hook window");
    ensure_cave_frame();
    int n = (int)plugin_blobs.size();
    uint64_t target_dst_val = 0;
    if (branch_host) {
        override_original = true;
        for (size_t i = 0; i < original.size(); i += 4) {
            uint32_t insn;
            memcpy(&insn, original.data() + i, 4);
            uint64_t tgt = target_insn(insn, hook_va + i);
            if (tgt) {
                target_dst_val = tgt;
                break;
            }
        }
        if (!target_dst_val)
            throw std::runtime_error("branch hook needs a branch instruction");
    }
    const int call_slot_size = (int)armcave::aarch64::kMaxBranchSequenceBytes;
    const int save_size = (int)g_cave_frame.save.size();
    const int restore_size = (int)g_cave_frame.restore.size();
    const int control_size = save_size + n * call_slot_size + restore_size +
        (override_original ? (int)g_cave_frame.ret.size() :
         (int)armcave::aarch64::max_relocated_size(original.size()) +
         (int)armcave::aarch64::kMaxBranchSequenceBytes);
    uint64_t calls_va = cave_va + save_size;
    uint64_t resume_va = cave_va + save_size + n * call_slot_size + restore_size;
    std::vector<uint8_t> patched_original;
    if (!override_original)
        patched_original = patched(original, hook_va, resume_va);
    std::vector<int> offsets;
    struct Chunk { int offset; std::vector<uint8_t> bytes; };
    std::vector<Chunk> chunks;
    int cur = control_size;
    for (int i = 0; i < n; i++) {
        auto *blob = plugin_blobs[i];
        cur = (cur + 3) & ~3;
        auto &regs = blob->register_args;
        int entry = blob->entry_offset;
        if (!regs.empty()) {
            int woff = cur;
            cur += plugin_wrapper_max_size(regs);
            cur = (cur + 3) & ~3;
            int poff = cur;
            offsets.push_back(woff);
            int text_aligned = (blob->max_text_bytes() + 15) & ~15;
            auto built = blob->build(cave_va + poff,
                cave_va + poff + text_aligned,
                target_binary);
            chunks.push_back({woff, make_wrapper(regs, cave_va + woff,
                                                 cave_va + poff + entry)});
            int built_size = (int)built.size();
            chunks.push_back({poff, std::move(built)});
            cur = poff + built_size;
        } else {
            offsets.push_back(cur + entry);
            int text_aligned = (blob->max_text_bytes() + 15) & ~15;
            auto built = blob->build(cave_va + cur,
                cave_va + cur + text_aligned,
                target_binary);
            int built_size = (int)built.size();
            chunks.push_back({cur, std::move(built)});
            cur += built_size;
        }
    }
    std::vector<uint8_t> out;
    out.insert(out.end(), g_cave_frame.save.begin(), g_cave_frame.save.end());
    for (int i = 0; i < n; i++) {
        auto call = armcave::aarch64::make_branch_sequence(
            calls_va + (uint64_t)i * call_slot_size,
            cave_va + offsets[i], true);
        out.insert(out.end(), call.begin(), call.end());
        while (out.size() < save_size + (size_t)(i + 1) * call_slot_size)
            out.insert(out.end(), NOP, NOP + 4);
    }
    out.insert(out.end(), g_cave_frame.restore.begin(), g_cave_frame.restore.end());
    if (override_original) {
        out.insert(out.end(), g_cave_frame.ret.begin(), g_cave_frame.ret.end());
    } else {
        out.insert(out.end(), patched_original.begin(), patched_original.end());
        auto branch = armcave::aarch64::make_branch_sequence(
            cave_va + out.size(), hook_va + hook_size, false);
        out.insert(out.end(), branch.begin(), branch.end());
    }
    if ((int)out.size() > control_size)
        throw std::runtime_error("hook dispatcher exceeded reserved size");
    out.resize((size_t)cur, 0);
    for (auto &chunk : chunks) {
        if (chunk.offset < 0 || (size_t)chunk.offset + chunk.bytes.size() > out.size())
            throw std::runtime_error("hook cave layout overflow");
        std::copy(chunk.bytes.begin(), chunk.bytes.end(), out.begin() + chunk.offset);
    }
    if (target_dst_val) {
        struct { const uint8_t *marker; uint64_t target; } markers[] = {
            {DST, target_dst_val},
            {NEXT, hook_va + 4},
            {CONV, hook_va + (uint64_t)hook_size}
        };
        for (auto &m : markers) {
            size_t idx = 0;
            while (true) {
                auto pos = out.begin() + (int)idx;
                auto found = std::search(pos, out.end(), m.marker, m.marker + 4);
                if (found == out.end()) break;
                int off = (int)(found - out.begin());
                auto branch = armcave::aarch64::make_branch_sequence(
                    cave_va + off, m.target, false);
                if (branch.size() != 4)
                    throw std::runtime_error("branch host marker is too small for far target");
                memcpy(&out[off], branch.data(), branch.size());
                idx = off + 1;
            }
        }
    }

    return out;
}

static void codesign(const std::filesystem::path &path) {
#ifdef __APPLE__
    std::string quoted = "'";
    for (char c : path.string()) {
        if (c == '\'') quoted += "'\\''";
        else quoted += c;
    }
    quoted += '\'';
    std::string cmd = "codesign --force --sign - " + quoted + " >/dev/null 2>/dev/null";
    system(cmd.c_str());
#else
    (void)path;
#endif
}

static int va_to_offset(BinaryImage *binary, uint64_t va) {
    auto off = binary->virtual_address_to_offset(va);
    if (!off)
        throw std::runtime_error("cannot map VA");
    return (int)*off;
}

void patch_hook_window(const std::filesystem::path &binary_path,
                       const std::filesystem::path &output_path,
                       uint64_t src_va, int size, uint64_t dst_va) {
    (void)binary_path;
    auto binary = BinaryImage::parse(output_path);
    if (!binary)
        throw std::runtime_error("failed to parse " + output_path.string());

    int off = va_to_offset(binary.get(), src_va);
    auto branch = armcave::aarch64::make_branch_sequence(src_va, dst_va, false);
    auto data = read_file(output_path);

    if ((int)branch.size() > size || size % 4)
        throw std::runtime_error("hook window too small");

    if (off < 0 || (size_t)off + (size_t)size > data.size())
        throw std::runtime_error("hook window is outside the file");
    memcpy(&data[off], branch.data(), branch.size());
    for (int pos = (int)branch.size(); pos < size; pos += 4)
        memcpy(&data[off + pos], NOP, 4);

    write_file(output_path, data);
}

void patch_call_window(const std::filesystem::path &binary_path,
                       const std::filesystem::path &output_path,
                       uint64_t src_va, int size, uint64_t dst_va) {
    (void)binary_path;
    auto binary = BinaryImage::parse(output_path);
    if (!binary)
        throw std::runtime_error("failed to parse " + output_path.string());

    int off = va_to_offset(binary.get(), src_va);
    auto call = armcave::aarch64::make_branch_sequence(src_va, dst_va, true);
    auto data = read_file(output_path);
    if ((int)call.size() > size || size % 4)
        throw std::runtime_error("call window too small");
    if (off < 0 || (size_t)off + (size_t)size > data.size())
        throw std::runtime_error("call window is outside the file");
    memcpy(&data[off], call.data(), call.size());
    for (int pos = (int)call.size(); pos < size; pos += 4)
        memcpy(&data[off + pos], NOP, 4);
    write_file(output_path, data);
}

void patch_bytes_va(const std::filesystem::path &binary_path,
                    const std::filesystem::path &output_path,
                    uint64_t va, const std::vector<uint8_t> &payload) {
    (void)binary_path;
    auto binary = BinaryImage::parse(output_path);
    if (!binary)
        throw std::runtime_error("failed to parse " + output_path.string());

    int off = va_to_offset(binary.get(), va);
    auto data = read_file(output_path);
    if (off + (int)payload.size() > (int)data.size())
        throw std::runtime_error("patch out of range");
    memcpy(&data[off], payload.data(), payload.size());
    write_file(output_path, data);
}

std::pair<uint64_t, uint64_t> patch_hook_macho(
    const std::filesystem::path &binary_path,
    const std::filesystem::path &output_path,
    int hook_file_off, int hook_size,
    const std::vector<uint8_t> &original,
    const std::vector<PluginBlob *> &plugin_blobs,
    const std::string &seg_name_str,
    bool override_original, bool branch_host,
    const std::vector<uint64_t> *nop_addrs) {

    auto before = open_macho(binary_path.string());
    if (!before.bin)
        throw std::runtime_error("failed to parse " + binary_path.string());

    ensure_cave_frame();

    int n = (int)plugin_blobs.size();
    int control = (int)g_cave_frame.save.size() +
                  n * (int)armcave::aarch64::kMaxBranchSequenceBytes +
                  (int)g_cave_frame.restore.size() +
                  (override_original ? (int)g_cave_frame.ret.size() :
                   patched_size(original) +
                   (int)armcave::aarch64::kMaxBranchSequenceBytes);
    int size = control;
    for (auto *b : plugin_blobs) {
        size += b->total_bytes();
        size += plugin_wrapper_max_size(b->register_args) + 16;
    }
    if (size < 4) size = 4;

    SegmentPlan plan;
    plan.name = seg_name_str;
    plan.size = size;
    plan.content.resize(size, 0);
    add_segment(output_path, plan, output_path);

    auto after = open_macho(output_path.string());
    if (!after.bin)
        throw std::runtime_error("failed to parse " + output_path.string());

    uint64_t hook_va = remap_macho_offset_va(*before.bin, *after.bin, hook_file_off);
    uint64_t cave_va = seg_va(*after.bin, seg_name_str, size);
    int cave_off = segment_file_offset(*after.bin, seg_name_str);
    int actual_hook_size = hook_window_size(hook_va, cave_va);
    if (actual_hook_size > (int)original.size())
        throw std::runtime_error("Mach-O hook window is too small for far trampoline");
    hook_size = std::max(hook_size, actual_hook_size);

    auto blob = build_hook_cave(cave_va, hook_va, hook_size, original, plugin_blobs,
                                 &output_path, override_original, branch_host, nop_addrs);

    if (size < (int)blob.size())
        throw std::runtime_error("Mach-O cave exceeded reserved segment");
    blob.resize(size, 0);

    write_at_offset(output_path, cave_off, blob, size);
    patch_hook_window(output_path, output_path, hook_va, hook_size, cave_va);
    codesign(output_path);

    return {hook_va, cave_va};
}
